### PR TITLE
MNTOR-1103 - ADR 0000 use es compatible unit test frameworks

### PR DESCRIPTION
**Pull request by rhelmer**
# References: 
Jira: MNTOR-1103

# Description

Proposal on how we should use ES-compatible unit test frameworks

# Screenshot (if applicable)

Not applicable.

# How to test

Not applicable.

# Checklist (Definition of Done)
- [ ] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).
- [x] I've added or updated the relevant sections in readme and/or code comments
- [ ] Product Owner accepted the User Story (demo of functionality completed) or waived the privilege.
- [ ] All acceptance criteria are met.
- [x] Jira ticket has been updated (if needed) to match changes made during the development process.
- [ ] Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage.

### CODE REVIEW

File: docs/adr/0000-use-es-compatible-unit-test-frameworks.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+Monitor V1 uses the [Jest](https://jestjs.io/) for unit tests. Jest is both a test runner and also a [mocking](https://jestjs.io/docs/mock-functions) library.
+V1 currently has [~30% code coverage](https://coveralls.io/builds/55941931) over 24 tests, and uses Jest mocks in 11 of
+these to mock 8 modules.
+
+Jest does not fully support ES6+ modules for several reasons:
+
+* Per the spec, ES6+ does not allow so-called "monkey-patching" already-loaded modules
+* Jest expects to be able to "hoist" it's `jest.mock` commands before modules are loaded, which also isn't allowed in ES6+
+* Node still requires an experimental flag to load ES modules as of this writing
```
Comment (toufali) [inline docs/adr/0000-use-es-compatible-unit-test-frameworks.md:25]: I'm not sure this line about ESM experimental flag is true?  From this page: https://nodejs.org/docs/latest-v18.x/api/esm.html#modules-ecmascript-modules
>Node.js fully supports ECMAScript modules as they are currently specified and provides interoperability between them and its original module format, [CommonJS](https://nodejs.org/docs/latest-v18.x/api/modules.html).

File: docs/adr/0000-use-es-compatible-unit-test-frameworks.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+* Bad, because both the system under test and also unit test need changes
+  * the system under test needs to be rewritten to support being tested without mocks
+  * the unit tests need to be rewritten to use the new non-jest test runner
+
+### Use an ES6+ compatible test runner, along with an ES6+ compatible mocking library
+
+* Good, because the system under test and the test code are pure ES modules
+* Good, because the system under test needs no changes
+* Bad, because the unit tests need to be rewritten to use the new non-jest test runner and mocking library
+* Bad, because an experimental feature is required (Node experimental module loader)
```
Comment (toufali) [inline docs/adr/0000-use-es-compatible-unit-test-frameworks.md:111]: As above, maybe I'm confusing the feature being referenced – I understood ESM is fully supported by our version of Node?

**Review by toufali — COMMENTED** [review]
Thanks for putting this down!  Minor question about experimental ESM in Node, otherwise all makes sense to me.

File: docs/adr/0000-use-es-compatible-unit-test-frameworks.md (hunk 1 above)
Comment (rhelmer) [reply docs/adr/0000-use-es-compatible-unit-test-frameworks.md:25]: Jest needs this specifically: `--experimental-vm-modules` https://nodejs.org/api/vm.html#vm_class_vm_module

File: docs/adr/0000-use-es-compatible-unit-test-frameworks.md (hunk 2 above)
Comment (rhelmer) [reply docs/adr/0000-use-es-compatible-unit-test-frameworks.md:111]: Sorry this is incorrect, will fix!

Comment (rhelmer) [reply docs/adr/0000-use-es-compatible-unit-test-frameworks.md:111]: Oops wait this is not incorrect - this is actually a different feature :)
`--loader` https://nodejs.org/api/esm.html#loaders

File: docs/adr/0000-use-es-compatible-unit-test-frameworks.md (hunk 1 above)
Comment (rhelmer) [reply docs/adr/0000-use-es-compatible-unit-test-frameworks.md]: OK I clarified which feature this is referring to, it's really that Jest (when running in pure ES mode) requires an experimenta

File: docs/adr/0000-use-es-compatible-unit-test-frameworks.md (hunk 2 above)
Comment (rhelmer) [reply docs/adr/0000-use-es-compatible-unit-test-frameworks.md:112]: I've clarified this as well (thanks for pointing it out, I didn't remember this detail at first when you pointed it out!)

**Pull request merged**